java-checks/src/main/java/org/sonar/java/checks/NullShouldNotBeUsedWithOptionalCheck.java

```diff
@@ -0,0 +1,134 @@
+  
+  private static final String NULLABLE = "javax.annotation.Nullable";
+
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethod(MethodTree method) {
+    if (returnsOptional(method)) {
+
+      // check that the method is not annotated with @Nullable
+      ModifiersTree modifiers = method.modifiers();
+      for (AnnotationTree annotation : modifiers.annotations()) {
+        Type type = annotation.annotationType().symbolType();
+        if (type.is(NULLABLE)) {
+          context.reportIssue(this, annotation, "Methods with an \"Optional\" return type should not be \"@Nullable\".");
+        }
+      }
+
+      // check that the method does not return "null"
+      method.accept(new ReturnVisitor());
+    }
+
+    super.visitMethod(method);
+  }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree binaryExpression) {
+    String message = "Remove this null-check of an \"Optional\".";
+    if (isComparisonOperator(binaryExpression.operatorToken())) {
+      ExpressionTree left = binaryExpression.leftOperand();
+      ExpressionTree right = binaryExpression.rightOperand();
+      if (isOptional(left) && isNull(right) ||
+          isNull(left) && isOptional(right)) {
+        context.reportIssue(this, binaryExpression, message);
+      }
+    }
+
+    super.visitBinaryExpression(binaryExpression);
+  }
+
+  /**
+   * A visitor that raises an issue on any <code>return null</code> statement. 
+   */
+  private class ReturnVisitor extends BaseTreeVisitor {
+
+    @Override
+    public void visitReturnStatement(ReturnStatementTree returnStatement) {
+      ExpressionTree expression = returnStatement.expression();
+      if (expression.is(Kind.NULL_LITERAL)) {
+        context.reportIssue(NullShouldNotBeUsedWithOptionalCheck.this, expression.firstToken(), "Methods with an \"Optional\" return type should never return null.");
+      }
+    }
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      // don't visit subclasstree, methods in there will be visited by outer class
+    }
+
+  }
+
+  /**
+   * Returns <code>true</code> iff the specified method signature returns an <code>Optional</code>.
+   */
+  private static boolean returnsOptional(MethodTree method) {
+    TypeTree returnType = method.returnType();
+    if (returnType != null) {
+      return returnType.symbolType().is(Optional.class.getName());
+    }
+    return false;
+  }
+
+  private static boolean isComparisonOperator(SyntaxToken token) {
+    return JavaPunctuator.EQUAL.getValue().equals(token.text()) ||
+           JavaPunctuator.NOTEQUAL.getValue().equals(token.text());
+  }
+
+  private static boolean isOptional(ExpressionTree expression) {
+    // warning: "null" is an Optional!
```

Comment:
No need for the warning. The `<nullType>` always return `true` when checking if it is another type. It is an expected behavor!